packages/palette/src/elements/Modal/ModalBase.tsx

```diff
@@ -0,0 +1,167 @@
+import { Flex, FlexProps } from "../Flex"
+
+// TODO: Update TypeScript definitions for this library
+// https://github.com/theKashey/react-remove-scroll
+const ScrollIsolation = styled(RemoveScroll as any)`
+  width: 100%;
+  height: 100%;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+`
+
+const Container = styled(Flex)`
+  position: fixed;
+  top: 0;
+  right: 0;
+  bottom: 0;
+  left: 0;
+  align-items: center;
+  justify-content: center;
+  ${systemZIndex}
+`
+
+const Dialog = styled(Flex).attrs({ role: "dialog" })`
+  overflow-y: auto;
+  -webkit-overflow-scrolling: touch;
+
+  > * {
+    max-height: 100%;
+  }
+`
+
+/**
+ * BaseModal
+ */
+export type ModalBaseProps = FlexProps &
+  ZIndexProps & {
+    children?: React.ReactNode
+    dialogProps?: FlexProps
+    onClose?(): void
+  }
+
+/**
+ * It seems we've landed on this value as the 'top'
+ */
+export const DEFAULT_MODAL_Z_INDEX = 9999
+
+const FOCUSABLE_SELECTOR = [
+  "a[href]",
+  "area[href]",
+  "input:not([disabled])",
+  "select:not([disabled])",
+  "textarea:not([disabled])",
+  "button:not([disabled])",
+  '[tabindex="0"]',
+].join(", ")
+
+/**
+ * BaseModal
+ * Low-level modal that has no opinions about layout/overlay
+ * Modals content using a portal, locks scroll.
+ */
+export const ModalBase: React.FC<ModalBaseProps> = ({
+  children,
+  zIndex = DEFAULT_MODAL_Z_INDEX,
+  dialogProps = {},
+  onClose = () => null,
+  ...rest
+}) => {
+  const appendEl = useRef(document.createElement("div"))
+  const containerEl = useRef<HTMLDivElement | null>(null)
+  const scrollIsolationEl = useRef<HTMLDivElement | null>(null)
+
+  const [focusableEls, setFocusableEls] = useState<HTMLElement[]>([])
+  const { index: focusableIndex, handlePrev, handleNext } = useCursor({
+    max: focusableEls.length,
+  })
+
+  useEffect(() => {
+    if (!focusableEls.length) return
+    focusableEls[focusableIndex].focus()
+  }, [focusableEls, focusableIndex])
+
+  const handleClick = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
+    if (event.target === scrollIsolationEl.current) {
+      onClose()
+    }
+  }
+
+  const handleKeydown = (event: KeyboardEvent) => {
+    switch (event.key) {
+      case "Escape":
+        // Prevent <esc> from interfering with the returned focus
+        event.preventDefault()
+        event.stopPropagation()
+
+        // Handle close
+        return onClose()
+
+      case "Tab":
+        // Lock focus within modal
+        event.preventDefault()
+        event.stopPropagation()
+
+        // Move focus up or down
+        event.shiftKey ? handlePrev() : handleNext()
+        break
+      default:
+        break
+    }
+  }
+
+  useEffect(() => {
+    const { current } = appendEl
+
+    const focusedElBeforeOpen = document.activeElement as HTMLElement
+
+    // Append the dialog
+    document.body.appendChild(current)
+
+    // Gets the focusable elements
+    const __focusableEls__ = [].slice.call(
```

Comment:
Out of curiosity what does the `__` signify?

Reply:
And can this use [Array.from](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from)?

Reply:
I should probably just change this to a single leading underscore which is probably more conventional. Just "temporary/private", since I don't want to wait for the callback on the setState to get at the actual els for the immediate focus handling.

Reply:
Ah yeah, I couldn't think of that method off the top of my head 😆